### PR TITLE
✨ Add support for Trio via AnyIO

### LINKED ISSUE #310: Trio -  Python library for async concurrency and I/O
**Description**

Hello!
I'm using Hypercorn with uvloop.
I'd like to use Hypercorn with trio, but the program doesn't start when i switch uvloop by trio.

Do you have plans to implement trio?

Thank's!

PS: Nice work.

### PR DESCRIPTION
**Pull request by graingert**
Fixes #310 

see also https://github.com/encode/starlette/pull/1157

### CODE REVIEW

File: fastapi/concurrency.py
Comment (ArcLightSlavik) [inline fastapi/concurrency.py:5]: Would need to flagged as a breaking change in the change log? (message for tiangolo he makes the change logs)

**Review by ArcLightSlavik — APPROVED** [review]

File: pyproject.toml
Comment (Kludex) [inline pyproject.toml:66]: It's an `starlette` requirement, so I guess we don't need it here?

File: fastapi/concurrency.py
Comment (graingert) [reply fastapi/concurrency.py]: starlette walked back on the removal of this function

File: pyproject.toml
Comment (graingert) [reply pyproject.toml]: this depends on `anyio[trio]` now, which isn't a starlette requirement

**Comment by graingert** [comment]
currently blocked on https://github.com/encode/starlette/pull/1210

**Comment by graingert** [comment]
@tiangolo can you enable my CI builds?

File: pyproject.toml
Comment (graingert) [inline pyproject.toml:35]:
```suggestion
    "starlette ==0.16.0",
```

File: pyproject.toml
Comment (graingert) [inline pyproject.toml:65]:
```suggestion
    "anyio[trio] >=3.2.1,<4",
```

**Comment by espetro** [comment]
Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FasAPI. How are we supposed to enable it?

**Comment by graingert** [comment]
> Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FastAPI. How are we supposed to enable it?

You'll need to pip install this branch from git

**Comment by ArcLightSlavik** [comment]
> > Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FastAPI. How are we supposed to enable it?
> 
> You'll need to pip install this branch from git

Antony to the rescue 😉 https://www.youtube.com/watch?v=AQrskWh-F5E

**Comment by dhirschfeld** [comment]
> You'll need to pip install this branch from git

As a `conda` user it would be great to have this in a release (`conda-forge` only package pypi releases).
@graingert has been keeping this up-to-date with `master` - is there anything stopping it being merged?

**Comment by adriangb** [comment]
> is there anything stopping it being merged?

I think the only thing missing is review from @tiangolo

**Comment by graingert** [comment]
> I also pinned Starlette to the minimum version that supports AnyIO (`0.15.0`) to make it easier for others to gradually migrate.

You should definitely pin to 0.16.0, the breaking change is around the TestClient(async_backend) kwarg which didn't exist before 0.15.0

**Pull request merged**
